programs/drift_vaults/src/state/vault.rs

```diff
@@ -504,4 +738,391 @@ impl Vault {
+                    vault_shares_after,
+                    manager_profit_share: 0,
+                    management_fee: management_fee_payment,
+                    management_fee_shares,
+                },
+                Some(VaultDepositorRecordProtocolParams {
+                    protocol_profit_share: 0,
+                    protocol_fee: protocol_fee_payment,
+                    protocol_fee_shares,
+                }),
+            )?;
+        }
+
+        Ok(())
+    }
+
+    pub fn protocol_withdraw(
+        &mut self,
+        vault_protocol: &mut Option<RefMut<VaultProtocol>>,
+        vault_equity: u64,
+        now: i64,
+    ) -> Result<u64> {
+        if vault_protocol.is_none() {
+            validate!(
+                false,
+                ErrorCode::VaultProtocolMissing,
+                "Protocol cannot withdraw for a non-protocol vault"
+            )?;
+        }
+
+        self.last_manager_withdraw_request
+            .check_redeem_period_finished(self, now)?;
+
+        self.apply_rebase(vault_protocol, vault_equity)?;
+
+        let VaultFee {
+            management_fee_payment,
+            management_fee_shares,
+            protocol_fee_payment,
+            protocol_fee_shares,
+        } = self.apply_fee(vault_protocol, vault_equity, now)?;
+
+        let vault_shares_before: u128 = self.get_protocol_shares(vault_protocol);
+        let total_vault_shares_before = self.total_shares;
+        let user_vault_shares_before = self.user_shares;
+
+        let n_shares = match vault_protocol {
+            None => 0,
+            Some(vp) => vp.last_protocol_withdraw_request.shares,
+        };
+
+        validate!(
+            n_shares > 0,
+            ErrorCode::InvalidVaultWithdraw,
+            "Must submit withdraw request and wait the redeem_period ({} seconds)",
+            self.redeem_period
+        )?;
+
+        let amount: u64 =
+            depositor_shares_to_vault_amount(n_shares, self.total_shares, vault_equity)?;
+
+        let n_tokens = match vault_protocol {
+            None => amount,
+            Some(vp) => amount.min(vp.last_protocol_withdraw_request.value),
+        };
+
+        validate!(
+            vault_shares_before >= n_shares,
+            ErrorCode::InsufficientVaultShares
+        )?;
+
+        self.total_withdraws = self.total_withdraws.saturating_add(n_tokens);
+        if let Some(vp) = vault_protocol {
+            vp.protocol_total_withdraws = vp.protocol_total_withdraws.saturating_add(n_tokens);
+        }
+        self.net_deposits = self.net_deposits.safe_sub(n_tokens.cast()?)?;
+
+        let vault_shares_before = self.get_protocol_shares(vault_protocol);
+
+        validate!(
+            vault_shares_before >= n_shares,
+            ErrorCode::InvalidVaultWithdrawSize,
+            "vault_shares_before={} < n_shares={}",
+            vault_shares_before,
+            n_shares
+        )?;
+
+        self.total_shares = self.total_shares.safe_sub(n_shares)?;
```

Comment:
should sub from `vp.protocol_profit_and_fee_shares` here

Reply:
oh i missed that sub below, but i dont think you should subtract from `total_shares` here since that would reduce the manager's share on `protocol_withdraw`

Reply:
not a bug, resolving